src/ucp/tag/rndv.c

```diff
@@ -5,19 +5,316 @@
  */
 
 #include "rndv.h"
+#include <ucp/proto/proto_am.inl>
+#include <ucp/core/ucp_request.inl>
+#include <ucs/datastruct/queue.h>
 
+#define UCP_ALIGN 256
+#define UCP_MTU_SIZE 4096
+
+static size_t ucp_tag_rndv_rts_pack(void *dest, void *arg)
+{
+    ucp_rndv_rts_hdr_t *rndv_rts_hdr = dest;
+    ucp_request_t *req = arg;
+
+    rndv_rts_hdr->super.tag       = req->send.tag;
+    rndv_rts_hdr->req.reqptr      = (uintptr_t)req;
+    rndv_rts_hdr->req.sender_uuid = req->send.ep->worker->uuid;
+    rndv_rts_hdr->address         = (uintptr_t)req->send.buffer;
+    rndv_rts_hdr->size            = req->send.length;
+    if (req->send.state.dt.contig.memh != UCT_INVALID_MEM_HANDLE) {
+        uct_md_mkey_pack(ucp_ep_md(req->send.ep, ucp_ep_get_rndv_data_lane(req->send.ep)),
+                         req->send.state.dt.contig.memh,
+                         rndv_rts_hdr + 1);
+        return sizeof(*rndv_rts_hdr) +
+                      ucp_ep_md_attr(req->send.ep,
+                                     ucp_ep_get_rndv_data_lane(req->send.ep))->rkey_packed_size;
+    } else {
+        /* For rndv emulation based on send-recv */
+        return sizeof(*rndv_rts_hdr);
+    }
+}
+
+static ucs_status_t ucp_proto_progress_rndv_rts(uct_pending_req_t *self)
+{
+    /* send the RTS. the pack_cb will pack all the necessary fields in the RTS */
+    ucs_status_t status = ucp_do_am_bcopy_single(self, UCP_AM_ID_RNDV_RTS,
+                                                 ucp_tag_rndv_rts_pack);
+    return status;
+}
 
 ucs_status_t ucp_tag_send_start_rndv(ucp_request_t *req)
 {
-    /* TODO */
-    ucs_fatal("unimplemented");
-    return UCS_ERR_NOT_IMPLEMENTED;
+    ucs_status_t status;
+
+    /* zcopy */
+    status = ucp_request_send_buffer_reg(req, ucp_ep_get_rndv_data_lane(req->send.ep));
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    ucp_ep_connect_remote(req->send.ep);
+
+    req->send.uct.func = ucp_proto_progress_rndv_rts;
+    return UCS_OK;
+}
+
+ucs_status_t ucp_proto_progress_rndv_get(uct_pending_req_t *self)
+{
+    ucp_request_t *get_req = ucs_container_of(self, ucp_request_t, send.uct);
+    ucs_status_t status;
+    size_t offset, length;
+
+    if (ucp_ep_is_stub(get_req->send.ep)) {
+        return UCS_ERR_NO_RESOURCE;
+    }
+
+    ucs_debug("send rndv get");
+
+    /* reset the lane to rndv since it might have been set to 0 since it was stub on RTS receive */
+    get_req->send.lane = ucp_ep_get_rndv_data_lane(get_req->send.ep);
+
+    if (get_req->send.state.dt.contig.memh == UCT_INVALID_MEM_HANDLE) {
+        /* TODO Not all UCTs need registration on the recv side */
+        status = ucp_request_send_buffer_reg(get_req, get_req->send.lane);
+        ucs_assert_always(status == UCS_OK);
+
+        {
+            size_t max_get_zcopy = ucp_ep_config(get_req->send.ep)->max_rndv_get_zcopy;
+            size_t remainder = (uintptr_t) get_req->send.buffer % UCP_ALIGN;
+
+            if (remainder) {
```

Comment:
this part can be done outside the 'if', since it is not related to memory registration. after you pass the ucp_ep_is_stub check, the configuration is already correct.